trieste/models/interfaces.py

```diff
@@ -241,6 +264,113 @@ def log(self) -> None:
             with tf.name_scope(f"{i}"):
                 model.log()
 
+    def reparam_sampler(self, num_samples: int) -> ReparametrizationSampler:
+        """
+        Return a reparameterization sampler providing `num_samples` samples across
+        all the models in the model stack. This is currently only implemented for
+        stacks made from models that have a :class:`BatchReparametrizationSampler`
+        as their reparameterization sampler.
+
+        :param num_samples: The desired number of samples.
+        :return: The reparametrization sampler.
+        :raise NotImplementedError: If the models in the stack do not share the
+            same :meth:`reparam_sampler`.
+        """
+
+        sampler_types = [type(model.reparam_sampler(1)) for model in self._models]
+        unique_sampler_types = set(sampler_types)
+        if len(unique_sampler_types) > 1:
```

Comment:
I'm confused. Did you make this change then revert it? Happy to discuss if doesn't seem to work.

Also, my original suggestion isn't quite right. You should check for type equality, not instance, as the optimization won't work correctly if any of the samplers are subtypes of BatchReparameterizationSampler:

```python
if all(type(sampler) is BatchReparameterizationSampler for sampler in samplers):
```